spec/lib/synapse/haproxy_spec.rb

```diff
@@ -28,37 +33,165 @@ class MockWatcher; end;
+    backends = [{ 'host' => 'somehost', 'port' => 5555}]
+    allow(mockWatcher).to receive(:backends).and_return(backends)
+    allow(mockWatcher).to receive(:generator_config).and_return({
+      'haproxy' => {'port' => 2200, 'disabled' => true}
+    })
     mockWatcher
   end
 
+  describe 'validates arguments' do
+    it 'succeeds on minimal config' do
+      conf = {
+        'global' => [],
+        'defaults' => [],
+        'do_writes' => false,
+        'do_reloads' => false,
+        'do_socket' => false
+      }
+      Synapse::ConfigGenerator::Haproxy.new(conf)
+      expect{Synapse::ConfigGenerator::Haproxy.new(conf)}.not_to raise_error
+    end
+
+    it 'validates req_pairs' do
+      req_pairs = {
+        'do_writes' => 'config_file_path',
+        'do_socket' => 'socket_file_path',
+        'do_reloads' => 'reload_command'
+      }
+      valid_conf = {
+        'global' => [],
+        'defaults' => [],
+        'do_reloads' => false,
+        'do_socket' => false,
+        'do_writes' => false
+      }
+
+      req_pairs.each do |key, value|
+        conf = valid_conf.clone
+        conf[key] = true
+        expect{Synapse::ConfigGenerator::Haproxy.new(conf)}.
+          to raise_error(ArgumentError, "the `#{value}` option is required when `#{key}` is true")
+      end
+
+    end
+
+    it 'properly defaults do_writes, do_socket, do_reloads' do
+      conf = {
+        'global' => [],
+        'defaults' => [],
+        'config_file_path' => 'test_file',
+        'socket_file_path' => 'test_socket',
+        'reload_command' => 'test_reload'
+      }
+
+      expect{Synapse::ConfigGenerator::Haproxy.new(conf)}.not_to raise_error
+      haproxy = Synapse::ConfigGenerator::Haproxy.new(conf)
+      expect(haproxy.opts['do_writes']).to eql(true)
+      expect(haproxy.opts['do_socket']).to eql(true)
+      expect(haproxy.opts['do_reloads']).to eql(true)
+    end
+
+    it 'complains when req_pairs are not passed at all' do
+      conf = {
+        'global' => [],
+        'defaults' => [],
+      }
+      expect{Synapse::ConfigGenerator::Haproxy.new(conf)}.to raise_error(ArgumentError)
+    end
+  end
+
   describe '#name' do
     it 'returns haproxy' do
       expect(subject.name).to eq('haproxy')
     end
   end
 
+  describe 'disabled watcher' do
+    let(:watchers) { [mockwatcher, mockwatcher_disabled] }
+    let(:socket_file_path) { 'socket_file_path' }
+
+    before do
+      config['haproxy']['do_socket'] = true
+      config['haproxy']['socket_file_path'] = socket_file_path
+    end
+
+    it 'does not generate config' do
+      allow(subject).to receive(:parse_watcher_config).and_return({})
+      expect(subject).to receive(:generate_frontend_stanza).exactly(:once).with(mockwatcher, nil)
+      expect(subject).to receive(:generate_backend_stanza).exactly(:once).with(mockwatcher, nil)
+      subject.update_config(watchers)
+    end
+
+    it 'does not cause a restart due to the socket' do
```

Comment:
Sure, I made it a context, except that we can't stub out talk_to_socket the same way for the two specs (one where the server was previously in the backends, and one without).

Also I don't like testing for falsey when that's not something that is supposed to be ok, namely nil isn't a valid value at all for restart_required.